main.go

```diff
@@ -80,94 +143,204 @@ func main() {
+	)
+
+	if c.CertFile != "" && c.KeyFile != "" {
+		if cert, err = tls.LoadX509KeyPair(c.CertFile, c.KeyFile); err != nil {
+			return nil, fmt.Errorf("failed to read certificate and private key %v", err)
 		}
 
-		if _, err = fmt.Fprint(writer, prettyJSON.String()); err != nil {
-			log.Printf("error while writting successful body response: %v\n", err)
-			return
+		certs = append(certs, cert)
+	}
+
+	if pool, err = x509.SystemCertPool(); err != nil {
+		return nil, fmt.Errorf("failed to read system root CAs %v", err)
+	}
+
+	if c.RootCA != "" {
+		if data, err = os.ReadFile(c.RootCA); err != nil {
+			return nil, fmt.Errorf("failed to read http client root ca: %w", err)
 		}
+
+		pool.AppendCertsFromPEM(data)
+	}
+
+	return &http.Client{
+		Timeout: c.Timeout,
+		Transport: &http.Transport{
+			Proxy: http.ProxyFromEnvironment,
+			DialContext: (&net.Dialer{
+				Timeout:   30 * time.Second,
+				KeepAlive: 30 * time.Second,
+				DualStack: true,
+			}).DialContext,
+			MaxIdleConns:          100,
+			ResponseHeaderTimeout: c.Timeout,
+			IdleConnTimeout:       90 * time.Second,
+			TLSHandshakeTimeout:   10 * time.Second,
+			ExpectContinueTimeout: 1 * time.Second,
+			MaxIdleConnsPerHost:   runtime.GOMAXPROCS(0) + 1,
+			TLSClientConfig: &tls.Config{
+				RootCAs:            pool,
+				MinVersion:         tls.VersionTLS12,
+				Certificates:       certs,
+				InsecureSkipVerify: true,
+			},
+		},
+	}, nil
+}
+
+func newHTTPRequest(clientID string) (req *http.Request, err error) {
+	if req, err = http.NewRequest("GET", fmt.Sprintf("%s?client_id=%s", resourceURL, clientID), nil); err != nil {
+		return nil, err
+	}
+
+	req.Header = http.Header{
+		"Content-Type":  []string{"application/json"},
+		"Authorization": []string{fmt.Sprintf("Bearer %s", token.AccessToken)},
+	}
+
+	if injectCertMode {
+		req.Header.Add("x-ssl-cert-hash", xsslCertHash)
+	}
+
+	return req, err
+}
+
+type WellKnownEndpoints struct {
+	Issuer                string              `json:"issuer"`
+	AuthorizationEndpoint string              `json:"authorization_endpoint"`
+	MtlsEndpointAliases   MtlsEndpointAliases `json:"mtls_endpoint_aliases"`
+}
+
+type MtlsEndpointAliases struct {
+	TokenEndpoint string `json:"token_endpoint"`
+}
+
+func (c *Config) fetchEndpointURLs() {
+	var (
+		resp *http.Response
+		we   WellKnownEndpoints
+		err  error
+	)
+
+	if resp, err = http.Get(c.WellKnownURL.String()); err != nil {
+		log.Fatalf("error retrieving .well-known %v", err)
+	}
+	defer resp.Body.Close()
+	json.NewDecoder(resp.Body).Decode(&we)
+
+	if c.IssuerURL, err = url.Parse(we.Issuer); err != nil {
+		log.Fatal("could not get /issure endpoint from .well-known")
+	}
+	if c.AuthorizeEndpoint, err = url.Parse(we.AuthorizationEndpoint); err != nil {
+		log.Fatal("could not get /authorize endpoint from .well-known")
+	}
+	if c.TokenEndpoint, err = url.Parse(we.MtlsEndpointAliases.TokenEndpoint); err != nil {
+		log.Fatal("could not get /token endpoint from .well-known")
 	}
```

Comment:
Is manual parsing necessary? Can't you just parse to a struct with *url.URL instead of string?
```
type WellKnownEndpoints struct {
	Issuer                *url.URL              `json:"issuer"`
	AuthorizationEndpoint *url.URL              `json:"authorization_endpoint"`
	MtlsEndpointAliases   MtlsEndpointAliases `json:"mtls_endpoint_aliases"`
}

type MtlsEndpointAliases struct {
	TokenEndpoint *url.URL `json:"token_endpoint"`
}
```

Reply:
As far as I can tell, that will fail with url.URL